## Rubric

Instructions: DELETE this cell before you submit via a `git push` to your repo before deadline. This cell is for your reference only and is not needed in your report. 

Scoring: Out of 10 points

- Each Developing  => -2 pts
- Each Unsatisfactory/Missing => -4 pts
  - until the score is 

If students address the detailed feedback in a future checkpoint they will earn these points back


|                  | Unsatisfactory                                                                                                                                                                                                    | Developing                                                                                                                                                                                              | Proficient                                     | Excellent                                                                                                                              |
|------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------|
| Data relevance   | Did not have data relevant to their question. Or the datasets don't work together because there is no way to line them up against each other. If there are multiple datasets, most of them have this trouble | Data was only tangentially relevant to the question or a bad proxy for the question. If there are multiple datasets, some of them may be irrelevant or can't be easily combined.                       | All data sources are relevant to the question. | Multiple data sources for each aspect of the project. It's clear how the data supports the needs of the project.                         |
| Data description | Dataset or its cleaning procedures are not described. If there are multiple datasets, most have this trouble                                                                                              | Data was not fully described. If there are multiple datasets, some of them are not fully described                                                                                                      | Data was fully described                       | The details of the data descriptions and perhaps some very basic EDA also make it clear how the data supports the needs of the project. |
| Data wrangling   | Did not obtain data. They did not clean/tidy the data they obtained.  If there are multiple datasets, most have this trouble                                                                                 | Data was partially cleaned or tidied. Perhaps you struggled to verify that the data was clean because they did not present it well. If there are multiple datasets, some have this trouble | The data is cleaned and tidied.                | The data is spotless and they used tools to visualize the data cleanliness and you were convinced at first glance                      |


# COGS 108 - Data Checkpoint

## Authors

Instructions: REPLACE the contents of this cell with your team list and their contributions. Note that this will change over the course of the checkpoints

This is a modified [CRediT taxonomy of contributions](https://credit.niso.org). For each group member please list how they contributed to this project using these terms:
> Analysis, Background research, Conceptualization, Data curation, Experimental investigation, Methodology, Project administration, Software, Visualization, Writing – original draft, Writing – review & editing

Example team list and credits:
- Alice Anderson: Conceptualization, Data curation, Methodology, Writing - original draft
- Bob Barker:  Analysis, Software, Visualization
- Charlie Chang: Project administration, Software, Writing - review & editing
- Dani Delgado: Analysis, Background research, Visualization, Writing - original draft

## Research Question

Instructions: REPLACE the contents of this cell with your work, including any updates to recover points lost in your proposal feedback



## Background and Prior Work

Instructions: REPLACE the contents of this cell with your work, including any updates to recover points lost in your proposal feedback

## Hypothesis


Instructions: REPLACE the contents of this cell with your work, including any updates to recover points lost in your proposal feedback


## Data

### Data overview

Instructions: REPLACE the contents of this cell with descriptions of your actual datasets.

For each dataset include the following information
- Dataset #1
  - Dataset Name:
  - Link to the dataset:
  - Number of observations:
  - Number of variables:
  - Description of the variables most relevant to this project
  - Descriptions of any shortcomings this dataset has with repsect to the project
- Dataset #2 (if you have more than one!)
  - same as above
- etc

Each dataset deserves either a set of bullet points as above or a few sentences if you prefer that method.

If you plan to use multiple datasets, add a few sentences about how you plan to combine these datasets.

In [19]:
# Run this code every time when you're actively developing modules in .py files.  It's not needed if you aren't making modules
#
## this code is necessary for making sure that any modules we load are updated here 
## when their source code .py files are modified

%load_ext autoreload
%autoreload 2

In [20]:
# Setup code -- this only needs to be run once after cloning the repo!
# this code downloads the data from its source to the `data/00-raw/` directory
# if the data hasn't updated you don't need to do this again!

# if you don't already have these packages (you should!) uncomment this line
# %pip install requests tqdm

import sys
sys.path.append('./modules') # this tells python where to look for modules to import

import get_data # this is where we get the function we need to download data

# replace the urls and filenames in this list with your actual datafiles
# yes you can use Google drive share links or whatever
# format is a list of dictionaries; 
# each dict has keys of 
#   'url' where the resource is located
#   'filename' for the local filename where it will be stored 
datafiles = [
    { 'url': 'https://raw.githubusercontent.com/fivethirtyeight/data/refs/heads/master/airline-safety/airline-safety.csv', 'filename':'airline-safety.csv'},
    { 'url': 'https://raw.githubusercontent.com/fivethirtyeight/data/refs/heads/master/bad-drivers/bad-drivers.csv', 'filename':'bad-drivers.csv'}
]

get_data.get_raw(datafiles,destination_directory='data/00-raw/')

Overall Download Progress:  50%|█████     | 1/2 [00:00<00:00,  6.59it/s]   

Successfully downloaded: airline-safety.csv



Overall Download Progress: 100%|██████████| 2/2 [00:00<00:00,  7.28it/s]

Successfully downloaded: bad-drivers.csv


### Dataset #1 – Portland Housing Market Data (2020–2021)

Dataset Description

This dataset contains housing market data for the Portland, Oregon area from 2020 to 2021. The data includes information such as:
	•	City and ZIP code
	•	Number of bedrooms and bathrooms
	•	Days on Zillow (time on market)
	•	Date sold
	•	Zestimate (Zillow’s estimated home value)

The primary variables used for this project are:
	•	zestimate: Estimated housing price (USD)
	•	daysOnZillow: Number of days the property remained on the market
	•	dateSold: Sale date (converted from Unix timestamp)
	•	year/month: Extracted from sale date for time analysis

This dataset allows us to analyze housing market trends before and after the implementation of Measure 110 in Oregon.


In [21]:
import pandas as pd

housing = pd.read_csv("data/00-raw/portland_housing.csv")

housing.head()

/tmp/ipykernel_414/3700533288.py:3: DtypeWarning: Columns (20,50,161,200) have mixed types. Specify dtype option on import or set low_memory=False.
  housing = pd.read_csv("data/00-raw/portland_housing.csv")


abbreviatedAddress address/city  address/zipcode  bathrooms  bedrooms  \
0  2860 NE Blossom Hill Rd     Fairview            97024        3.0       3.0   
1    20386 NE Mackenzie Ln     Fairview            97024        3.0       3.0   
2          1121 SW 10th Dr      Gresham            97080        3.0       4.0   
3       19309 NE Glisan St     Portland            97230        1.0       3.0   
4         25-65 NW 20th St      Gresham            97030        3.0       6.0   

                         brokerageName      dateSold  daysOnZillow  \
0                                  NaN  1.621560e+12          25.0   
1  Harcourts Real Estate Network Group  1.619140e+12          53.0   
2             ERA Freeman & Associates  1.622760e+12          11.0   
3         Premiere Property Group, LLC  1.622760e+12          11.0   
4                                  NaN  1.622510e+12          14.0   

                                         description  favoriteCount  ...  \
0  2860 NE Blossom Hill Rd, Fairview, OR 97024 is...            NaN  ...   
1  Cute as a button 3bed/2.5bath home with brand ...            NaN  ...   
2  What a wonderful large home at end of quiet cu...            NaN  ...   
3                                                NaN            1.0  ...   
4  25-65 NW 20th St, Gresham, OR 97030 is a apart...            NaN  ...   

  taxHistory/0/taxPaid taxHistory/0/time  taxHistory/0/value  \
0              2989.74      1.592260e+12            172290.0   
1                  NaN               NaN                 NaN   
2              5178.62      1.592260e+12            282340.0   
3              2420.36      1.497560e+12            142330.0   
4              2809.61      1.592260e+12            153180.0   

   taxHistory/0/valueIncreaseRate  \
0                        0.029950   
1                             NaN   
2                        0.029987   
3                        0.029959   
4                        0.029989   

                                                 url  yearBuilt  zestimate  \
0  https://www.zillow.com/homedetails/2860-NE-Blo...     2007.0   317406.0   
1  https://www.zillow.com/homedetails/20386-NE-Ma...     2001.0        NaN   
2  https://www.zillow.com/homedetails/1121-SW-10t...     1982.0   512355.0   
3  https://www.zillow.com/homedetails/19309-NE-Gl...     1953.0        NaN   
4  https://www.zillow.com/homedetails/25-65-NW-20...     1967.0        NaN   

   zestimateHighPercent zestimateLowPercent        zpid  
0                   5.0                 5.0    72258538  
1                   NaN                 NaN  2073593774  
2                   5.0                 5.0    53959026  
3                   NaN                 NaN   176563865  
4                   NaN                 NaN    53819568  

[5 rows x 348 columns]

Cleaning choices:
I reduced the raw Zillow dataset (25731 rows, 348 columns) to a smaller set of relevant variables for my research question: city/zipcode, bedrooms, bathrooms, daysOnZillow, dateSold, and zestimate. Many priceHistory/ and taxHistory/ columns contained substantial missingness and were not needed for our analysis, so I excluded them. I converted dateSold to datetime and coerced invalid entries to missing.

Data Cleaning and Preparation

The dataset originally contained over 300 variables. We reduced the dataset to the most relevant variables for analysis. The following cleaning steps were performed:
	•	Converted Unix timestamp in dateSold to datetime format
	•	Extracted year and month
	•	Converted numerical columns to proper numeric types
	•	Checked missing values
	•	Filtered the data to include only 2020–2021 observations
	•	Saved the cleaned dataset to data/02-processed

In [22]:
housing.shape

(25731, 348)

In [23]:
housing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25731 entries, 0 to 25730
Columns: 348 entries, abbreviatedAddress to zpid
dtypes: bool(11), float64(111), int64(6), object(220)
memory usage: 66.4+ MB


In [24]:
housing.isnull().sum().sort_values(ascending=False).head(10)

resoFacts/securityFeatures/0                   20752
priceHistory/10/attributeSource/infoString2    20576
resoFacts/interiorFeatures/11                  20537
priceHistory/10/price                          20509
priceHistory/10/showCountyLink                 20485
priceHistory/10/time                           20485
priceHistory/10/date                           20485
priceHistory/10/event                          20485
priceHistory/10/postingIsRental                20485
priceHistory/10/source                         20485
dtype: int64

In [25]:
housing.columns

Index(['abbreviatedAddress', 'address/city', 'address/zipcode', 'bathrooms',
       'bedrooms', 'brokerageName', 'dateSold', 'daysOnZillow', 'description',
       'favoriteCount',
       ...
       'taxHistory/0/taxPaid', 'taxHistory/0/time', 'taxHistory/0/value',
       'taxHistory/0/valueIncreaseRate', 'url', 'yearBuilt', 'zestimate',
       'zestimateHighPercent', 'zestimateLowPercent', 'zpid'],
      dtype='object', length=348)

In [26]:
keep_cols = [
    "address/city",
    "address/zipcode",
    "bedrooms",
    "bathrooms",
    "daysOnZillow",
    "dateSold",
    "zestimate",
]

housing_small = housing[keep_cols].copy()

housing_small.head()

address/city  address/zipcode  bedrooms  bathrooms  daysOnZillow  \
0     Fairview            97024       3.0        3.0          25.0   
1     Fairview            97024       3.0        3.0          53.0   
2      Gresham            97080       4.0        3.0          11.0   
3     Portland            97230       3.0        1.0          11.0   
4      Gresham            97030       6.0        3.0          14.0   

       dateSold  zestimate  
0  1.621560e+12   317406.0  
1  1.619140e+12        NaN  
2  1.622760e+12   512355.0  
3  1.622760e+12        NaN  
4  1.622510e+12        NaN

In [27]:
housing_small.isnull().sum().sort_values(ascending=False)

zestimate          1974
bedrooms            770
bathrooms           484
daysOnZillow          6
address/city          0
address/zipcode       0
dateSold              0
dtype: int64

In [28]:
housing_small["dateSold"] = pd.to_datetime(
    housing_small["dateSold"],
    unit="ms",
    errors="coerce"
)

In [29]:
housing_small["year"] = housing_small["dateSold"].dt.year
housing_small["month"] = housing_small["dateSold"].dt.month

housing_small[["dateSold", "year", "month"]].head()

dateSold  year  month
0 2021-05-21 01:20:00  2021      5
1 2021-04-23 01:06:40  2021      4
2 2021-06-03 22:40:00  2021      6
3 2021-06-03 22:40:00  2021      6
4 2021-06-01 01:13:20  2021      6

In [31]:
housing_small.describe()

address/zipcode      bedrooms     bathrooms  daysOnZillow  \
count     25731.000000  24961.000000  25247.000000  25725.000000   
mean      97143.830943      3.356356      2.561896    184.738620   
min        7123.000000      0.000000      0.000000      1.000000   
25%       97062.000000      3.000000      2.000000     77.000000   
50%       97203.000000      3.000000      3.000000    185.000000   
75%       97221.000000      4.000000      3.000000    292.000000   
max       97267.000000     43.000000     23.000000    422.000000   
std         567.759775      1.001626      1.009135    113.431466   

                            dateSold     zestimate          year         month  
count                          25731  2.375700e+04  25731.000000  25731.000000  
mean   2021-01-16 10:11:43.540476416  6.441574e+05   2020.524659      6.731064  
min              2020-06-12 00:00:00  9.170000e+04   2020.000000      1.000000  
25%              2020-10-02 00:53:20  4.495000e+05   2020.000000      4.000000  
50%              2021-01-15 00:20:00  5.737000e+05   2021.000000      7.000000  
75%              2021-05-04 23:13:20  7.259000e+05   2021.000000      9.000000  
max              2021-07-15 23:00:00  6.537500e+06   2021.000000     12.000000  
std                              NaN  3.500689e+05      0.499401      3.106898

In [32]:
housing_small.groupby("year")["zestimate"].mean()

year
2020    645255.774941
2021    643123.900735
Name: zestimate, dtype: float64

In [33]:
housing_filtered = housing_small[
    (housing_small["year"] >= 2020) &
    (housing_small["year"] <= 2021)
].copy()

housing_filtered.shape

(25731, 9)

In [34]:
import os
os.makedirs("data/02-processed", exist_ok=True)

housing_filtered.to_csv(
    "data/02-processed/housing_clean.csv",
    index=False
)

### Dataset #2 

See instructions above for Dataset #1.  Feel free to keep adding as many more datasets as you need.  Put each new dataset in its own section just like these. 

Lastly if you do have multiple datasets, add another section where you demonstrate how you will join, align, cross-reference or whatever to combine data from the different datasets

Please note that you can always keep adding more datasets in the future if these datasets you turn in for the checkpoint aren't sufficient.  The goal here is demonstrate that you can obtain and wrangle data.  You are not tied down to only use what you turn in right now.

In [36]:
import os
os.listdir("data/00-raw")

['VSRR_Provisional_Drug_Overdose_Death_Counts_20260218.csv',
 '.gitkeep',
 'portland_housing.csv',
 'airline-safety.csv',
 'bad-drivers.csv']

In [37]:
import pandas as pd

overdose = pd.read_csv("data/00-raw/VSRR_Provisional_Drug_Overdose_Death_Counts_20260218.csv")

overdose.head()

State  Year     Month           Period        Indicator Data Value  \
0    AK  2015   January  12 month-ending  Cocaine (T40.5)        NaN   
1    AK  2015  February  12 month-ending  Cocaine (T40.5)        NaN   
2    AK  2015     March  12 month-ending  Cocaine (T40.5)        NaN   
3    AK  2015     April  12 month-ending  Cocaine (T40.5)        NaN   
4    AK  2015       May  12 month-ending  Cocaine (T40.5)        NaN   

   Percent Complete  Percent Pending Investigation State Name  \
0               100                            0.0     Alaska   
1               100                            0.0     Alaska   
2               100                            0.0     Alaska   
3               100                            0.0     Alaska   
4               100                            0.0     Alaska   

                                            Footnote Footnote Symbol  \
0  Numbers may differ from published reports usin...              **   
1  Numbers may differ from published reports usin...              **   
2  Numbers may differ from published reports usin...              **   
3  Numbers may differ from published reports usin...              **   
4  Numbers may differ from published reports usin...              **   

  Predicted Value  
0             NaN  
1             NaN  
2             NaN  
3             NaN  
4             NaN

In [39]:
overdose["State Name"].unique()

array(['Alaska', 'Alabama', 'Arkansas', 'Arizona', 'California',
       'Colorado', 'Connecticut', 'District of Columbia', 'Delaware',
       'Florida', 'Georgia', 'Hawaii', 'Iowa', 'Idaho', 'Illinois',
       'Indiana', 'Kansas', 'Kentucky', 'Louisiana', 'Massachusetts',
       'Maryland', 'Maine', 'Michigan', 'Minnesota', 'Missouri',
       'Mississippi', 'Montana', 'North Carolina', 'North Dakota',
       'Nebraska', 'New Hampshire', 'New Jersey', 'New Mexico', 'Nevada',
       'New York', 'Ohio', 'Oklahoma', 'Oregon', 'Pennsylvania',
       'Puerto Rico', 'Rhode Island', 'South Carolina', 'South Dakota',
       'Tennessee', 'Texas', 'United States', 'Utah', 'Virginia',
       'Vermont', 'Washington', 'Wisconsin', 'West Virginia', 'Wyoming',
       'New York City'], dtype=object)

In [40]:
overdose_or = overdose[overdose["State Name"] == "Oregon"].copy()

overdose_or.head()

State  Year     Month           Period        Indicator Data Value  \
56115    OR  2015   January  12 month-ending  Cocaine (T40.5)         20   
56116    OR  2015  February  12 month-ending  Cocaine (T40.5)         19   
56117    OR  2015     March  12 month-ending  Cocaine (T40.5)         18   
56118    OR  2015     April  12 month-ending  Cocaine (T40.5)         18   
56119    OR  2015       May  12 month-ending  Cocaine (T40.5)         14   

       Percent Complete  Percent Pending Investigation State Name  \
56115               100                       0.023248     Oregon   
56116               100                       0.023091     Oregon   
56117               100                       0.023006     Oregon   
56118               100                       0.020047     Oregon   
56119               100                       0.019910     Oregon   

                                                Footnote Footnote Symbol  \
56115  Numbers may differ from published reports usin...              **   
56116  Numbers may differ from published reports usin...              **   
56117  Numbers may differ from published reports usin...              **   
56118  Numbers may differ from published reports usin...              **   
56119  Numbers may differ from published reports usin...              **   

      Predicted Value  
56115              20  
56116              19  
56117              18  
56118              18  
56119              14

In [41]:
overdose_or = overdose_or[
    (overdose_or["Year"] >= 2020) &
    (overdose_or["Year"] <= 2021)
].copy()

overdose_or.shape

(288, 12)

In [42]:
overdose_or["Indicator"].unique()

array(['Cocaine (T40.5)', 'Heroin (T40.1)', 'Methadone (T40.3)',
       'Natural & semi-synthetic opioids (T40.2)',
       'Natural & semi-synthetic opioids, incl. methadone (T40.2, T40.3)',
       'Natural, semi-synthetic, & synthetic opioids, incl. methadone (T40.2-T40.4)',
       'Number of Deaths', 'Number of Drug Overdose Deaths',
       'Opioids (T40.0-T40.4,T40.6)', 'Percent with drugs specified',
       'Psychostimulants with abuse potential (T43.6)',
       'Synthetic opioids, excl. methadone (T40.4)'], dtype=object)

In [44]:
overdose_or[overdose_or["Indicator"].str.contains("Opioid", case=False, na=False)]["Indicator"].unique()

array(['Natural & semi-synthetic opioids (T40.2)',
       'Natural & semi-synthetic opioids, incl. methadone (T40.2, T40.3)',
       'Natural, semi-synthetic, & synthetic opioids, incl. methadone (T40.2-T40.4)',
       'Opioids (T40.0-T40.4,T40.6)',
       'Synthetic opioids, excl. methadone (T40.4)'], dtype=object)

In [45]:
overdose_opioid = overdose_or[
    overdose_or["Indicator"] == "Opioids (T40.0-T40.4, T40.6)"
].copy()

overdose_opioid.head()

Empty DataFrame
Columns: [State, Year, Month, Period, Indicator, Data Value, Percent Complete, Percent Pending Investigation, State Name, Footnote, Footnote Symbol, Predicted Value]
Index: []

In [46]:
overdose_opioid["date"] = pd.to_datetime(
    overdose_opioid["Year"].astype(str) + "-" + overdose_opioid["Month"],
    errors="coerce"
)

In [47]:
overdose_opioid["Data Value"] = pd.to_numeric(
    overdose_opioid["Data Value"],
    errors="coerce"
)

In [48]:
overdose_monthly = overdose_opioid.groupby(
    ["Year", "Month"]
)["Data Value"].mean().reset_index()

overdose_monthly.head()

Empty DataFrame
Columns: [Year, Month, Data Value]
Index: []

In [49]:
import os
os.makedirs("data/02-processed", exist_ok=True)

overdose_monthly.to_csv(
    "data/02-processed/overdose_clean.csv",
    index=False
)

## Ethics

Our project uses public datasets on housing prices and drug overdose deaths.
These datasets do not contain personal identity information (PII), which reduces privacy concerns.

However, there are still moral considerations:

1. Data representation and bias
The housing dataset reflects the listed properties and may not fully represent all housing market transactions. Similarly, there may also be a delay in reporting data on drug overdose
Or the classification among the states is inconsistent.

2. Sensitive topics
Drug overdose deaths are a sensitive public health issue. We must be careful not to oversimplify the conclusion or imply causality; only correlation exists.

3. Be responsible for explanation
Any relationship between housing trends and the rate of drug overdose will be interpreted with caution. Correlation does not imply causation or external socio-economic factors
It may affect two trends.
Our goal is to objectively analyze the data and responsibly present the results.

## Team Expectations 

We expect all team members to contribute equally to data cleaning, analysis, and writing.
Each member will:

- Participate in weekly discussions
- Complete assigned tasks on time
- Communicate progress clearly
- Help debug and review each other's work

If conflicts arise, we will discuss them openly and resolve them respectfully.
All members are responsible for ensuring the final submission meets course standards.


## Project Timeline Proposal

Week 1
- Finalize research question
- Identify and download relevant datasets

Week 2
- Clean and preprocess housing and overdose datasets
- Perform exploratory data analysis (EDA)

Week 3
- Merge datasets and conduct statistical analysis
- Generate visualizations

Week 4
- Interpret results
- Draft final report

Final Week
- Revise report
- Finalize visualizations
- Submit project